include/nn/hpad/hpad.h

```diff
@@ -0,0 +1,312 @@
+   HPAD_SUBSTICK_EMULATION_LEFT  = 0x00010000,
+   HPAD_SUBSTICK_EMULATION_RIGHT = 0x00020000,
+   HPAD_SUBSTICK_EMULATION_DOWN  = 0x00040000,
+   HPAD_SUBSTICK_EMULATION_UP    = 0x00080000,
+} HPADButtons;
+
+typedef enum HPADMotorCommand
+{
+   HPAD_MOTOR_COMMAND_STOP      = 0,
+   HPAD_MOTOR_COMMAND_RUMBLE    = 1,
+   HPAD_MOTOR_COMMAND_STOP_HARD = 2,
+} HPADMotorCommand;
+
+struct HPADStatus
+{
+   //! Indicates what HPADButtons are held down.
+   int32_t hold;
+   //! Indicates what HPADButtons have been pressed since last sample.
+   int32_t trigger;
+   //! Indicates what HPADButtons have been released since last sample.
+   int32_t release;
+   //! X-Position of the analog stick
+   int8_t stickX;
+   //! Y-Position of the analog stick
+   int8_t stickY;
+   //! X-Position of the analog substick/C-Stick
+   int8_t substickX;
+   //! Y-Position of the analog substick/C-Stick
+   int8_t substickY;
+   //! Position of the left analog trigger
+   uint8_t triggerL;
+   //! Position of the right analog trigger
+   uint8_t triggerR;
+   //! Status of the controller
+   uint8_t status;
+   WUT_PADDING_BYTES(0x1);
+   //! Read error, 0 for success
+   int32_t error;
+};
+WUT_CHECK_OFFSET(HPADStatus, 0x00, hold);
+WUT_CHECK_OFFSET(HPADStatus, 0x04, trigger);
+WUT_CHECK_OFFSET(HPADStatus, 0x08, release);
+WUT_CHECK_OFFSET(HPADStatus, 0x0C, stickX);
+WUT_CHECK_OFFSET(HPADStatus, 0x0D, stickY);
+WUT_CHECK_OFFSET(HPADStatus, 0x0E, substickX);
+WUT_CHECK_OFFSET(HPADStatus, 0x0F, substickY);
+WUT_CHECK_OFFSET(HPADStatus, 0x10, triggerL);
+WUT_CHECK_OFFSET(HPADStatus, 0x11, triggerR);
+WUT_CHECK_OFFSET(HPADStatus, 0x12, status);
+WUT_CHECK_OFFSET(HPADStatus, 0x14, error);
+WUT_CHECK_SIZE(HPADStatus, 0x18);
+
+struct HPADGGGGStatus
+{
+   //! \c TRUE if the adapeter is connected
+   BOOL connected;
+   //! \c TRUE if the power supply is connected
+   BOOL powerSupplyConnected;
+   //! \c TRUE if the adapter is active/is sending samples
+   BOOL active;
+};
+WUT_CHECK_OFFSET(HPADGGGGStatus, 0x0, connected);
+WUT_CHECK_OFFSET(HPADGGGGStatus, 0x4, powerSupplyConnected);
+WUT_CHECK_OFFSET(HPADGGGGStatus, 0x8, active);
+WUT_CHECK_SIZE(HPADGGGGStatus, 0xC);
+
+typedef void (*HPADConnectCallback)(HPADChan chan, int32_t status);
+typedef void (*HPADGGGGConnectCallback)(HPADGGGGChan chan, BOOL connected);
+typedef void (*HPADPowerSupplyCallback)(HPADGGGGChan chan, BOOL connected);
+typedef void (*HPADSamplingCallback)(HPADGGGGChan chan);
+typedef void (*HPADResetCallback)(HPADGGGGChan chan, int32_t status);
+
+/**
+ * Initializes the HPAD library.
+ *
+ * \return
+ * 0 on success.
+ */
+int32_t
+HPADInit(void);
+
+/**
+ * Shuts down the HPAD library.
+ *
+ * \return
+ * 0 on success.
+ */
+int32_t
+HPADShutdown(void);
+
+/**
+ * Reads status buffers from a specified HPAD channel.
+ * 
+ * \warning
+ * The count parameter is ignored and this function will always fill up to 16 sampling buffers.
+ * Make sure to always make space for 16 buffers.
```

Comment:
Not sure what the best way of handling this would be. A warning like this might be enough or a wrapper struct/typedef containing 16 buffers could be created.

Reply:
Wrapper that either makes passing count != 16 an error or buffers appropriately would be my thought.